File: JAVA/Herança.java
abstract class Animal {
    abstract String som();
}

class Cachorro extends Animal {
    @Override
    String som() {
        return "Au Au!";
    }
}

class Gato extends Animal {
    @Override
    String som() {
        return "Miau!";
    }
}

public class Main {
    public static void main(String[] args) {
        Animal cachorro = new Cachorro();
        Animal gato = new Gato();

        System.out.println("Cachorro: " + cachorro.som());
        System.out.println("Gato: " + gato.som());
    }
}

File: GOLANG/Herança.go
package main

import "fmt"

type Animal interface {
    Som() string
}

type Cachorro struct{}

func (c Cachorro) Som() string {
    return "Au Au!"
}

type Gato struct{}

func (g Gato) Som() string {
    return "Miau!"
}

func main() {
    var animais []Animal
    animais = append(animais, Cachorro{})
    animais = append(animais, Gato{})

    for _, animal := range animais {
        fmt.Println(animal.Som())
    }
}
